File: perf/app/src/main/java/com/google/firebase/quickstart/perfmon/java/MainActivity.java
package com.google.firebase.quickstart.perfmon.java;

import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.firebase.perf.FirebasePerformance;
import com.google.firebase.perf.metrics.Trace;
import com.google.firebase.quickstart.perfmon.R;
import com.google.firebase.quickstart.perfmon.databinding.ActivityMainBinding;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Random;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    private static final String DEFAULT_CONTENT_FILE = "default_content.txt";
    private static final String CONTENT_FILE = "content.txt";
    private static final String IMAGE_URL =
            "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png";

    private Trace mTrace;

    private String STARTUP_TRACE_NAME = "startup_trace";
    private String REQUESTS_COUNTER_NAME = "requests sent";
    private String FILE_SIZE_COUNTER_NAME = "file size";
    private CountDownLatch mNumStartupTasks = new CountDownLatch(2);

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // write 40 chars of random text to file
                File contentFile = new File(MainActivity.this.getFilesDir(), CONTENT_FILE);

                writeStringToFile(contentFile.getAbsolutePath(), getRandomString(40) + "\n")
                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                if (!task.isSuccessful()) {
                                    Log.e(TAG, "Unable to write to file", task.getException());
                                    return;
                                }

                                loadFileFromDisk();
                            }
                        });
                }
        });

        // Begin tracing app startup tasks.
        mTrace = FirebasePerformance.getInstance().newTrace(STARTUP_TRACE_NAME);
        Log.d(TAG, "Starting trace");
        mTrace.start();
        loadImageFromWeb();
        // Increment the counter of number of requests sent in the trace.
        Log.d(TAG, "Incrementing number of requests counter in trace");
        mTrace.incrementMetric(REQUESTS_COUNTER_NAME, 1);
        loadFileFromDisk();
        // Wait for app startup tasks to complete asynchronously and stop the trace.
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    mNumStartupTasks.await();
                } catch (InterruptedException e) {
                    Log.e(TAG, "Unable to wait for startup task completion.");
                } finally {
                    Log.d(TAG, "Stopping trace");
                    mTrace.stop();
                    MainActivity.this.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Toast.makeText(MainActivity.this, "Trace completed",
                                    Toast.LENGTH_SHORT).show();
                        }
                    });
                }
            }
        }).start();
    }

    private void loadImageFromWeb() {
        Glide.with(this).
                load(IMAGE_URL)
                .placeholder(new ColorDrawable(ContextCompat.getColor(this, R.color.colorAccent)))
                .listener(new RequestListener<Drawable>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        mNumStartupTasks.countDown();
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        mNumStartupTasks.countDown();
                        return false;
                    }
                }).into(binding.headerIcon);
    }

    private Task<Void> writeStringToFile(final String filename, final String content) {
        final TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();
        Executors.newSingleThreadExecutor().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    FileOutputStream fos = new FileOutputStream(filename, true);
                    fos.write(content.getBytes());
                    fos.close();
                    taskCompletionSource.setResult(null);
                } catch (IOException e) {
                    e.printStackTrace();
                }

            }
        });
        return taskCompletionSource.getTask();
    }

    private Task<String> loadStringFromFile() {
        final TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
        Executors.newSingleThreadExecutor().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    File contentFile = new File(getFilesDir(), CONTENT_FILE);
                    if (contentFile.createNewFile()) {
                        // Content file exist did not exist in internal storage and new file was created.
                        // Copy in the default content.
                        InputStream is;
                        is = getAssets().open(DEFAULT_CONTENT_FILE);
                        int size = is.available();
                        byte[] buffer = new byte[size];
                        is.read(buffer);
                        is.close();
                        FileOutputStream fos = new FileOutputStream(contentFile);
                        fos.write(buffer);
                        fos.close();
                    }
                    FileInputStream fis = new FileInputStream(contentFile);
                    byte[] content = new byte[(int) contentFile.length()];
                    fis.read(content);
                    taskCompletionSource.setResult(new String(content));
                } catch (IOException e) {
                    e.printStackTrace();
                }

            }
        });
        return taskCompletionSource.getTask();
    }

    private void loadFileFromDisk() {
        loadStringFromFile()
                .addOnCompleteListener(new OnCompleteListener<String>() {
                    @Override
                    public void onComplete(@NonNull Task<String> task) {
                        if (!task.isSuccessful()) {
                            Log.e(TAG, "Couldn't read text file.");
                            Toast.makeText(
                                    MainActivity.this, getString(R.string.text_read_error), Toast.LENGTH_LONG)
                                    .show();
                            return;
                        }

                        String fileContent = task.getResult();
                        binding.textViewContent.setText(task.getResult());
                        // Increment a counter with the file size that was read.
                        Log.d(TAG, "Incrementing file size counter in trace");
                        mTrace.incrementMetric(FILE_SIZE_COUNTER_NAME, fileContent.getBytes().length);
                        mNumStartupTasks.countDown();
                    }
                });
    }

    private String getRandomString(int length) {
        char[] chars = "abcdefghijklmnopqrstuvwxyz".toCharArray();
        StringBuilder sb = new StringBuilder();
        Random random = new Random();
        for (int i = 0; i < length; i++) {
            char c = chars[random.nextInt(chars.length)];
            sb.append(c);
        }
        return sb.toString();
    }
}

File: perf/app/src/main/java/com/google/firebase/quickstart/perfmon/kotlin/MainActivity.kt
package com.google.firebase.quickstart.perfmon.kotlin

import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.core.content.ContextCompat
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.firebase.ktx.Firebase
import com.google.firebase.perf.ktx.performance
import com.google.firebase.perf.metrics.Trace
import com.google.firebase.quickstart.perfmon.R
import com.google.firebase.quickstart.perfmon.databinding.ActivityMainBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.util.Random
import java.util.concurrent.CountDownLatch

class MainActivity : AppCompatActivity() {

    private lateinit var trace: Trace

    private val numStartupTasks = CountDownLatch(2)

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.button.setOnClickListener {
            // write 40 chars of random text to file
            val contentFile = File(this.filesDir, CONTENT_FILE)

            writeStringToFile(contentFile.absolutePath, "${getRandomString(40)}\n")
                    .addOnCompleteListener { task ->
                        if (!task.isSuccessful) {
                            Log.e(TAG, "Unable to write to file", task.exception)
                            return@addOnCompleteListener
                        }

                        loadFileFromDisk()
                    }
        }

        // Begin tracing app startup tasks.
        trace = Firebase.performance.newTrace(STARTUP_TRACE_NAME)
        Log.d(TAG, "Starting trace")
        trace.start()
        loadImageFromWeb()
        // Increment the counter of number of requests sent in the trace.
        Log.d(TAG, "Incrementing number of requests counter in trace")
        trace.incrementMetric(REQUESTS_COUNTER_NAME, 1)
        loadFileFromDisk()
        // Wait for app startup tasks to complete asynchronously and stop the trace.
        Thread(Runnable {
            try {
                numStartupTasks.await()
            } catch (e: InterruptedException) {
                Log.e(TAG, "Unable to wait for startup task completion.")
            } finally {
                Log.d(TAG, "Stopping trace")
                trace.stop()
                runOnUiThread {
                    Toast.makeText(this, "Trace completed",
                            Toast.LENGTH_SHORT).show()
                }
            }
        }).start()
    }

    private fun loadImageFromWeb() {
        Glide.with(this).load(IMAGE_URL)
            .placeholder(ColorDrawable(ContextCompat.getColor(this, R.color.colorAccent)))
            .listener(object : RequestListener<Drawable> {
                override fun onLoadFailed(
                    e: GlideException?,
                    model: Any?,
                    target: Target<Drawable>?,
                    isFirstResource: Boolean
                ): Boolean {
                    numStartupTasks.countDown() // Signal end of image load task.
                    return false
                }

                override fun onResourceReady(
                    resource: Drawable?,
                    model: Any?,
                    target: Target<Drawable>?,
                    dataSource: DataSource?,
                    isFirstResource: Boolean
                ): Boolean {
                    numStartupTasks.countDown() // Signal end of image load task.
                    return false
                }
            }).into(binding.headerIcon)
    }

    private fun writeStringToFile(filename: String, content: String): Task<Void> {
        val taskCompletionSource = TaskCompletionSource<Void>()
        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                val fos = FileOutputStream(filename, true)
                fos.write(content.toByteArray())
                fos.close()
                taskCompletionSource.setResult(null)
            }
        }
        return taskCompletionSource.task
    }

    private fun loadStringFromFile(): Task<String> {
        val taskCompletionSource = TaskCompletionSource<String>()
        lifecycleScope.launch {
            withContext(Dispatchers.IO) {
                val contentFile = File(filesDir, CONTENT_FILE)
                if (contentFile.createNewFile()) {
                    // Content file exist did not exist in internal storage and new file was created.
                    // Copy in the default content.
                    val `is`: InputStream = assets.open(DEFAULT_CONTENT_FILE)
                    val size = `is`.available()
                    val buffer = ByteArray(size)
                    `is`.read(buffer)
                    `is`.close()
                    val fos = FileOutputStream(contentFile)
                    fos.write(buffer)
                    fos.close()
                }
                val fis = FileInputStream(contentFile)
                val content = ByteArray(contentFile.length().toInt())
                fis.read(content)
                taskCompletionSource.setResult(String(content))
            }
        }
        return taskCompletionSource.task
    }

    private fun loadFileFromDisk() {
        loadStringFromFile()
                .addOnCompleteListener(OnCompleteListener { task ->
                    if (!task.isSuccessful) {
                        Log.e(TAG, "Couldn't read text file.")
                        Toast.makeText(
                                this, getString(R.string.text_read_error), Toast.LENGTH_LONG)
                                .show()
                        return@OnCompleteListener
                    }

                    val fileContent = task.result
                    binding.textViewContent.text = task.result
                    // Increment a counter with the file size that was read.
                    Log.d(TAG, "Incrementing file size counter in trace")
                    trace.incrementMetric(
                            FILE_SIZE_COUNTER_NAME,
                            fileContent!!.toByteArray().size.toLong())
                    numStartupTasks.countDown()
                })
    }

    private fun getRandomString(length: Int): String {
        val chars = "abcdefghijklmnopqrstuvwxyz".toCharArray()
        val sb = StringBuilder()
        val random = Random()
        for (i in 0 until length) {
            val c = chars[random.nextInt(chars.size)]
            sb.append(c)
        }
        return sb.toString()
    }

    companion object {
        private const val TAG = "MainActivity"

        private const val DEFAULT_CONTENT_FILE = "default_content.txt"
        private const val CONTENT_FILE = "content.txt"
        private const val IMAGE_URL =
                "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png"

        private const val STARTUP_TRACE_NAME = "startup_trace"
        private const val REQUESTS_COUNTER_NAME = "requests sent"
        private const val FILE_SIZE_COUNTER_NAME = "file size"
    }
}
